### PR TITLE
Keyword searching with Azul (SCP-3806)

### PR DESCRIPTION
**Pull request by bistline**
This update extends keyword-only and keyword + faceted searching to HCA projects in the Azul data service.  Results are weighted/highlighted as normal using existing functionality.  Since Azul does not natively support keyword searching at the current time, keywords are "converted" into a faceted search by leveraging the work done in #1231. As we now have a complete list of all possible facet/filters from Azul, we can take any user-supplied keyword and attempt to map it to a known filter in Azul.  If a partial match is found, all possible filter values are taken and then used as the search criteria.  Results are then filtered to include only those where the supplied keyword is present in the project name/description.  This is to maintain parity with SCP-based keyword search functionality, and avoid confusion for users who would not expect to get a keyword match on an entry that does not contain the value in the name/description.  Downstream work will be done in SCP-3815 to log these "misses" to Mixpanel to inform us if these should be included in future download results.

MANUAL TESTING
1. Pull this branch, and run the following command in the Rails console to ensure that all results from Azul are stored in your local instance, as work on other branches may have cleared these values out.  This is not an issue for deployed environments as previous migrations have already seeded the data:
```
SearchFacet.update_all_facet_filters
```
2. Boot as normal, and sign in with an admin account that has `cross_dataset_search_backend` enabled
3. Enter `pulmonary` as a search term, and validate the following 3 HCA projects are returned
Screen Shot 2021-11-10 at 12 23 24 PM
4. Select `10X 5' v2 sequencing` from `library_preparation_protocol` (note the case of capital `X` and `sequencing` at the end) and ensure only one HCA project is returned:
Screen Shot 2021-11-10 at 12 28 36 PM
 
This PR satisfies SCP-3806

### CODE REVIEW

**Comment by bistline** [comment]
Build failure is the same as what @ehanna4 observed earlier.  Marking as a false positive.
```
  1) Failure:
FireCloudClientTest#test_create_and_manage_workspace [/home/app/webapp/test/integration/fire_cloud_client_test.rb:142]:
Did not receive correct confirmation, expected 'Your Google bucket fc-96207bf0-e46c-40bc-b7aa-41c6dea2dccd will be deleted within 24h.' but found 'You cannot delete this workspace: HttpResponse(202 Accepted,List(Date: Wed, 10 Nov 2021 18:00:15 GMT, Server: akka-http/10.2.6, X-Frame-Options: SAMEORIGIN, X-XSS-Protection: 1; mode=block, X-Content-Type-Options: nosniff, Strict-Transport-Security: max-age=31536000; includeSubDomains, Access-Control-Allow-Origin: *, Access-Control-Allow-Headers: authorization, content-type, accept, origin, x-app-id, Access-Control-Allow-Methods: GET, POST, PUT, PATCH, DELETE, OPTIONS, HEAD, Access-Control-Max-Age: 1728000, Via: 1.1 google, Alt-Svc: clear),HttpEntity.Chunked(text/plain; charset=UTF-8),HttpProtocol(HTTP/1.1))'
```

**Review by eweitz — APPROVED** [review]
Code looks good!  I suggested two maintainability refinements.

I recall you had performance concerns.  How many seconds does a keyword search take with Azul?  How many with the old approach?

File: app/lib/azul_search_service.rb [hunk 1]
```diff
@@ -11,8 +11,9 @@ class AzulSearchService
   # each Azul result entry under 'hits' will have these keys, whether project- or file-based
   RESULT_FACET_FIELDS = %w[samples specimens cellLines donorOrganisms organoids cellSuspensions].freeze
 
-  def self.append_results_to_studies(existing_studies, selected_facets:, terms:, facet_map: {})
-    azul_results = ::AzulSearchService.get_results(selected_facets: selected_facets)
+  def self.append_results_to_studies(existing_studies, selected_facets:, terms:, facet_map: nil)
+    facet_map ||= {}
```
[the remark this answers is not in the document]
Comment (bistline) [reply app/lib/azul_search_service.rb:15]: Ah yes - I can add that.  This has to do with behavior in Ruby that had me also scratching my head a bit.  Basically, setting a default value for the parameter `facet_map` does not work if `nil` is explicitly passed in.  Apparently, the "conventional" way of handling this is the `nil` default with the conditional assignment inside the method.  I can add a comment to reflect this.

File: app/lib/azul_search_service.rb [hunk 2]
```diff
@@ -22,11 +23,17 @@ def self.append_results_to_studies(existing_studies, selected_facets:, terms:, f
...
 
   # execute a search against Azul API
-  # TODO: add support for keyword search (SCP-3806)
-  def self.get_results(selected_facets:)
+  def self.get_results(selected_facets:, terms:)
     client = ApplicationController.hca_azul_client
     results = {}
-    query_json = client.format_query_from_facets(selected_facets)
+    facet_query = client.format_query_from_facets(selected_facets) if selected_facets
+    terms_to_facets = client.convert_keyword_to_facet_query(terms) if terms
```
[the remark this answers is not in the document]
Comment (bistline) [reply app/lib/azul_search_service.rb:30]: Good call - can do.

**Comment by bistline** [comment]
> Code looks good! I suggested two maintainability refinements.
> 
> I recall you had performance concerns. How many seconds does a keyword search take with Azul? How many with the old approach?

Since keyword queries are in effect faceted search requests now in Azul, the performance is the same for both - which is not great.  It's on the order of 3-4s, but this is on top of calling BigQuery, and happens on every pagination request, which makes each page take roughly 8s to load.  This has to do with the fact that we aren't paginating results from Azul, but instead asking for _everything_, and then doing the pagination on our end.  This means requests are needlessly loading all results, but then only returning a handful.  In addition, we're also querying for file access information (urls, sizes, names, etc.) based off of those results so that we can put information into the bulk download modal. This was to maintain parity with how searches in TDR worked, and was mostly to prevent test failures.  I learned later that Azul responses contain summary file information (types, sizes, and counts) that would be sufficient to populating the modal.  The plan is to refactor this in SCP-3812 so that we only request the access information when a bulk download request is issued.   I expect this will cut search requests down by roughly half, to around 2s.  Not great, but definitely a lot better.

**Review by devonbush — APPROVED** [review]

**Pull request merged**